Multisig/UI/Settings/OwnerKeyManagement/KeystoneOwnerKey/KeystoneSignFlow.swift

```diff
@@ -101,21 +101,35 @@ extension SECP256K1.UnmarshaledSignature {
         let s = data[32..<64]
         let v: UInt8
 
-        // only if v was overflown and is legacy transaction
-        if isLegacyTx && data.count > 65 {
+        let chainIdInt = UInt64(chainId) ?? 0
+        let vBytes: Bytes
 
+        // if v was overflown (e.g. chain_id > 109 according to EIP-155)
+        if data.count > 65 {
             // max 8 bytes to fit into UInt64
-            let vBytes: Bytes = [UInt8](data.suffix(from: 64).prefix(8))
+            let vBytes = [UInt8](data.suffix(from: 64).prefix(8))
             let vInt =  UInt64(vBytes)
-            let chainIdInt = UInt64(chainId) ?? 0
-
-            // recover V by deducting (chainId * 2 + 35) according to EIP-155.
+            // recover V by deducting (chainId * 2 + 35) according to EIP-155
             let vRecovered = vInt - (chainIdInt * 2 + 35)
             v = try! UInt8(vRecovered % 256)
         } else {
-            v = data[64]
+            vBytes = [UInt8]([data[64]])
+            let vInt = UInt8(vBytes)
+            if isLegacyTx {
+                // Legacy ethereum (pre-eip-155) adds 27 to v
+                v = vInt - 27
+            } else {
+                // v still can be chainId * 2 + 35 for non-legacy transactions (chaiId >=0)
+                if vInt >= 35 {
+                    let vRecovered = UInt64(vBytes) - (chainIdInt * 2 + 35)
+                    v = try! UInt8(vRecovered % 256)
+                } else {
+                    v = vInt
+                }
+            }
         }
-        assert(v == 0 || v == 1)
+        
+        assert(v == 0 || v == 1 || v == 27 || v == 28)
```

Comment:
the correct values for `v` are `0, 1, 2, 3, 27, 28, 29, 30`

Reply:
am I wrong about `2, 3, 29, 30`?

Reply:
According to the source code here

https://github.com/Boilertalk/secp256k1/blob/d5407179912e8c1f825a212a474aaa86b10f1352/src/ecdsa_impl.h#L292

the `recid` can have values of `0, 1, 2, and 2^1=3`, therefore v can have values `27, 28, 29, 30`. qed :D